### ISSUE TITLE
Incorrect higher order function return type monomorphisation

### ISSUE THREAD

**Issue by ehllie** [microsoft/pyright#4161]
**Describe the bug**
Pyright is not able to monomorphise the parameter types of generic functions returned from higher order generic functions.

**To Reproduce**
Type check the attached code with pyright

**Expected behavior**
Pyright correctly monomorphising the types of functions returned by `s_comb`, resulting in the types of `a`, `b` being inferred as `int` and `c`, `d` as `str`

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from itertools import tee
from typing import TYPE_CHECKING, Callable, Iterable, Iterator, TypeVar

from typing_extensions import reveal_type

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")


def s_comb(
    binary: Callable[[A, B], C],
    unary: Callable[[A], B],
    clone: Callable[[A], tuple[A, A]] = lambda x: (x, x),
) -> Callable[[A], C]:
    def comb(x: A) -> C:
        x1, x2 = clone(x)
        return binary(x1, unary(x2))

    return comb


def tail(iterable: Iterable[A]) -> Iterator[A]:
    iterator = iter(iterable)
    try:
        next(iterator)
    except StopIteration:
        ...
    yield from iterator


if __name__ == "__main__":
    print("Optional function provided:")
    for a, b in s_comb(zip, tail, tee)(n * 2 for n in range(20)):
        if TYPE_CHECKING:
            reveal_type(a)
            reveal_type(b)
        print(a, b)

    print("\nOptional function default:")
    for c, d in s_comb(zip, tail)("some string"):
        if TYPE_CHECKING:
            reveal_type(c)
            reveal_type(d)
        print(c, d)

```

**VS Code extension or command-line**
I am using pyright as a command line tool, and the command `pyright --version` returned `pyright 1.1.276`

**Additional context**
Output of type checking the provided code:
```console
No configuration file found.
No pyproject.toml file found.
stubPath /home/ellie/Code/type_bug/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.276
/home/ellie/Code/type_bug/parametric_polymorphism.py
  /home/ellie/Code/type_bug/parametric_polymorphism.py:36:25 - information: Type of "a" is "int"
  /home/ellie/Code/type_bug/parametric_polymorphism.py:37:25 - information: Type of "b" is "A@tail | int"
  /home/ellie/Code/type_bug/parametric_polymorphism.py:41:24 - error: Argument of type "Type[zip[Unknown]]" cannot be assigned to parameter "binary" of type "(A@s_comb, B@s_comb) -> C@s_comb" in function "s_comb"
    No overloaded function matches type "(Iterable[A@tail], Iterator[A@tail]) -> zip[tuple[_T1@__new__, A@tail]]"
      Type "(__iter1: Iterable[_T1@__new__], *, strict: bool = ...) -> zip[tuple[_T1@__new__]]" cannot be assigned to type "(Iterable[A@tail], Iterator[A@tail]) -> zip[tuple[_T1@__new__, A@tail]]"
        Function accepts too many positional parameters; expected 1 but received 2
          Function return type "zip[tuple[A@tail]]" is incompatible with type "zip[tuple[_T1@__new__, A@tail]]"
            "zip[tuple[A@tail]]" is incompatible with "zip[tuple[_T1@__new__, A@tail]]"
      Type "(__iter1: Iterable[_T1@__new__], __iter2: Iterable[_T2@__new__], *, strict: bool = ...) -> zip[tuple[_T1@__new__, _T2@__new__]]" cannot be assigned to type "(Iterable[A@tail], Iterator[A@tail]) -> zip[tuple[_T1@__new__, A@tail]]"
        Function return type "zip[tuple[A@tail, A@tail]]" is incompatible with type "zip[tuple[_T1@__new__, A@tail]]"
          "zip[tuple[A@tail, A@tail]]" is incompatible with "zip[tuple[_T1@__new__, A@tail]]"
    ... (reportGeneralTypeIssues)
  /home/ellie/Code/type_bug/parametric_polymorphism.py:41:17 - error: Argument of type "(x: A@s_comb) -> tuple[A@s_comb, A@s_comb]" cannot be assigned to parameter "clone" of type "(A@s_comb) -> tuple[A@s_comb, A@s_comb]" in function "s_comb"
    Type "(x: A@s_comb) -> tuple[A@s_comb, A@s_comb]" cannot be assigned to type "(A@s_comb) -> tuple[A@s_comb, A@s_comb]"
      Type "Iterable[A@tail]" cannot be assigned to type "A@s_comb"
      Function return type "tuple[A@s_comb, A@s_comb]" is incompatible with type "tuple[A@s_comb, A@s_comb]"
        "builtins.tuple" is incompatible with "builtins.tuple"
          Tuple entry 1 is incorrect type
            Type "A@s_comb" cannot be assigned to type "Iterable[A@tail]" (reportGeneralTypeIssues)
  /home/ellie/Code/type_bug/parametric_polymorphism.py:43:25 - information: Type of "c" is "_T1@__new__"
  /home/ellie/Code/type_bug/parametric_polymorphism.py:44:25 - information: Type of "d" is "str"
2 errors, 0 warnings, 4 informations
Completed in 0.809sec
```

**Comment by ehllie** [comment]
To expand a little more:
```py
if __name__ == "__main__":
    print("Optional function provided:")
    f1 = s_comb(zip, tail, tee)
    reveal_type(f1)
    for a, b in f1(n * 2 for n in range(20)):
        if TYPE_CHECKING:
            reveal_type(a)
            reveal_type(b)
        print(a, b)

    print("\nOptional function default:")
    f2 = s_comb(zip, tail)
    reveal_type(f2)
    for c, d in f2("some string"):
        if TYPE_CHECKING:
            reveal_type(c)
            reveal_type(d)
        print(c, d)
```
Pyright is almost able to correctly evaluate the type of `f1`:
`Pyright: Type of "f1" is "(Iterator[_T@tee]) -> zip[tuple[_T@tee, A@tail | _T@tee]]"`,
But it leaves the `A@tail` type in the signature, even though that type is the same as the type `_T@tee`

However the second function is the source of the actual errors despite seeming simpler:
`Pyright: Type of "f2" is "(Iterable[A@tail]) -> zip[tuple[_T1@__new__, A@tail]]"`

**Comment by erictraut** [comment]
There are two issues here.

First, there is not enough information available for pyright to analyze the lambda you're using for a default argument value in `s_comb`. In Python, there's no way to specify the input parameter types for a lambda, so these types must be inferred using bidirectional type inference. 

It is very unusual to specify a lambda in as a default argument, which is probably why no one has reported a similar issue in the past. In this particular case, there is insufficient information to infer the type of the lambda in a manner that allows for the type variables to be solved in the final call.

You can work around this limitation and eliminate the type error by replacing the lambda with a reference to an annotated function.

```python
def make_tuple(x: A) -> tuple[A, A]:
    return (x, x)

def s_comb(
    binary: Callable[[A, B], C],
    unary: Callable[[A], B],
    clone: Callable[[A], tuple[A, A]] = make_tuple,
) -> Callable[[A], C]:
    ...
```

As for the unresolved `A@tail` type variable, that is to be expected here. The problem is that type variables are "solved" in the context of a call to a generic function. You are not calling `tail`, so the type variable in its scope (`A@tail`) cannot be solved through a call to `s_comb`. You're asking too much of the current Python type system here.

I'll note that mypy is also unable to resolve the types in your code sample.

**Comment by ehllie** [comment]
I understand how you might not be interested in addressing this, but I'm gonna leave with a bit of an unfair comparison showing how this kind of feature works in Haskell:
```hs
main :: IO ()
main = do
  let pairwise = zip <*> tail
  forM_ (pairwise [1 .. 20]) \(a, b) -> do
    print [a, b]
  forM_ (pairwise "some string") \(c, d) -> do
    print [c, d]
```
And these are the types from the snippet:
```hs
pairwise :: [a] -> [(a, a)]
zip :: forall a b. [a] -> [b] -> [(a, b)]
<*> :: forall a b c. (a -> b -> c) -> (a -> b) -> a -> c
tail :: forall a. [a] -> [a]
```
The `<*>` (s_comb from my python code) signature is a lie, but in the snippet it's equivalent.

While I would not expect Python's type system to compete with Haskells, as those are vastly different languages, it would be a great feature to solve the generics of functions that are used as generic arguments,
